gk/main.c

```diff
@@ -16,31 +16,488 @@
+
+		struct {
+			uint8_t src[16];
+			uint8_t dst[16];
+		} v6;
+	} f;
+};
+
+/* Store information about a packet. */
+struct ipacket {
+	struct ip_flow  flow;
+	struct rte_mbuf *pkt;
+};
+
+struct flow_entry {
+	/* IP flow information. */
+	struct ip_flow flow;
+
+	/* The state of the entry. */
+	enum gk_flow_state state;
+
+	union {
+		struct {
+			/* The time the last packet of the entry was seen. */
+			uint64_t last_packet_seen_at;
+			/* 
+			 * The priority associated to
+			 * the last packet of the entry.
+			 */
+			uint8_t last_priority;
+			/* 
+			 * The number of packets that the entry is allowed
+			 * to send with @last_priority without waiting
+			 * the amount of time necessary to be granted
+			 * @last_priority.
+			 */
+			uint8_t allowance;
+			/* 
+			 * The ID of the Grantor server to which packets to
+			 * @dst must be sent.
+			 */
+			int grantor_id;
+		} request;
+
+		struct {
+			/* When the granted capability expires. */
+			uint64_t cap_expire_at;
+			/* When @budget_byte is reset. */
+			uint64_t budget_renew_at;
+			/* 
+			 * When @budget_byte is reset, reset it to
+			 * @tx_rate_kb_cycle * 1024 bytes.
+			 */
+			int tx_rate_kb_cycle;
+			/* How many bytes @src can still send in current cycle. */
+			int budget_byte;
+			/*
+			 * The ID of the Grantor server to which packets to
+			 * @dst must be sent.
+			 */
+			int grantor_id;
+			/* When GK should send the next renewal to @grantor_id. */
+			uint64_t send_next_renewal_at;
+			/*
+			 * How many cycles (unit) GK must wait before
+			 * sending the next capability renewal request.
+			 */
+			uint64_t renewal_step_cycle;
+		} granted;
+
+		struct {
+			/*
+			 * When the punishment (i.e. the declined capability)
+			 * expires.
+			 */
+			uint64_t expire_at;
+		} declined;
+	} u;
+};
+
+/* To support the optimized implementation of generic RSS hash function. */
+static uint8_t rss_key_be[RTE_DIM(default_rss_key)];
+
+/*
+ * XXX Optimized generic implementation of RSS hash function.
+ * If you want the calculated hash value matches NIC RSS value,
+ * you have to use special converted key with rte_convert_rss_key() fn.
+ * @param input_tuple
+ *   Pointer to input tuple with network order.
+ * @param input_len
+ *   Length of input_tuple in 4-bytes chunks.
+ * @param *rss_key
+ *   Pointer to RSS hash key.
+ * @return
+ *   Calculated hash value.
+ */
+static inline uint32_t
+gk_softrss_be(const uint32_t *input_tuple, uint32_t input_len,
+		const uint8_t *rss_key)
+{
+	uint32_t i;
+	uint32_t j;
+	uint32_t ret = 0;
+
+	for (j = 0; j < input_len; j++) {
+		/*
+		 * Need to use little endian,
+		 * since it takes ordering as little endian in both bytes and bits.
+		 */
+		uint32_t val = rte_be_to_cpu_32(input_tuple[j]);
+		for (i = 0; i < 32; i++)
+			if (val & (1 << (31 - i)))
+				ret ^= ((const uint32_t *)rss_key)[j] << i |
+					(uint32_t)((uint64_t)(((const uint32_t *)rss_key)[j + 1]) >> (32 - i));
+	}
+
+	return ret;
+}
+
+static inline uint32_t
```

Comment:
Drop the `inline` tag, not only is this function too long (it instantiates the inline function `gk_softrss_be()` twice), but it also is used as a pointer, so it cannot be inline. The compiler is already ignoring the tag due to the pointer case.